myhpom/forms/upload_requirements.py

```diff
@@ -37,13 +38,19 @@ def clean(self):
             name, extension = os.path.splitext(data['document'].name)
             data['document'].name = "%s-%s%s" % (name, str(uuid.uuid4())[:6], extension)
 
+            # (try to) generate a thumbnail of the document
+            try:
+                thumbnail_name = os.path.splitext(data['document'].name)[0] + '.jpg'
+                thumbnail_data = AdvanceDirective(document=data['document']).render_thumbnail_data()
+                data['thumbnail'] = SimpleUploadedFile(thumbnail_name, thumbnail_data)
+            except:
+                data['document'] = ''
+                self.add_error(
+                    'document',
+                    'The uploaded PDF file was corrupt or invalid and could not be saved.',
+                )
```

Comment:
A big issue I see with the strategy of validating on clean is that down the road we want to put thumbnail creation into a background task; I don't see an obvious way to break this out of this 'create AD' http request if thumbnails are generated here. Maybe that is okay at the moment, in the spirit of an MVP?

Since we already know we want to support another long running asynchronous task (third party service document verification) and probably add a status bit to track the state of this AD, we can refactor this chunk of code at that point.

I'm fine leaving this as is provided you add some comments that identify the issue, and its future resolution.

Reply:
Very good point. I've put in a big comment explaining the issue and likely future resolution, and will merge and deploy.